### PR TITLE
Make the battle timer tick every second

### PR DESCRIPTION
**Pull request by urkerab**
This might not be the best way of tracking whether the timer button exists...

### CODE REVIEW

File: js/client-battle.js [hunk 1]
```diff
@@ -421,8 +421,20 @@
 			}
 			return '<button name="openTimer" class="button timerbutton' + timerTicking + '"><i class="fa fa-hourglass-start"></i> ' + time + '</button>';
 		},
+		timerInterval: 0,
+		timerTick: function () {
+			if (typeof this.battle.kickingInactive == "number") this.battle.kickingInactive--;
```
Comment (Zarel) [inline js/client-battle.js]: Don't let it fall below `1`...

Comment (Zarel) [inline js/client-battle.js]: (But, otherwise, I approve of this method of detecting what state the timer is in)

File: js/client-battle.js [hunk 2]
```diff
@@ -421,8 +421,20 @@
...
+			this.updateTimer();
+		},
 		updateTimer: function () {
-			this.$('.timerbutton').replaceWith(this.getTimerHTML());
+			var timerButton = this.$('.timerbutton');
+			if (timerButton.length) {
+				timerButton.replaceWith(this.getTimerHTML());
+				if (!this.timerInterval) this.timerInterval = setInterval(this.timerTick.bind(this), 1000);
+			} else if (timerInterval) {
```
Comment (Zarel) [inline js/client-battle.js]: I'm assuming this is a typo of `this.timerInterval` and also the cause of the build error.

File: js/client-battle.js [hunk 3]
```diff
@@ -421,8 +421,20 @@
...
 			return '<button name="openTimer" class="button timerbutton' + timerTicking + '"><i class="fa fa-hourglass-start"></i> ' + time + '</button>';
 		},
+		timerInterval: 0,
+		timerTick: function () {
+			if (typeof this.battle.kickingInactive == "number") this.battle.kickingInactive--;
+			this.updateTimer();
+		},
 		updateTimer: function () {
-			this.$('.timerbutton').replaceWith(this.getTimerHTML());
+			var timerButton = this.$('.timerbutton');
```
Comment (Zarel) [inline js/client-battle.js]: `$timerButton` is the style for jQuery objects

File: js/client-battle.js [hunk 4]
```diff
@@ -421,8 +421,20 @@
...
+			if (timerButton.length) {
+				timerButton.replaceWith(this.getTimerHTML());
+				if (!this.timerInterval) this.timerInterval = setInterval(this.timerTick.bind(this), 1000);
```
Comment (Zarel) [inline js/client-battle.js]: Are you sure you want to be setting this in `updateTimer`? You should most likely be doing this in `getTimerHTML` itself. `updateTimer` is only called for specific updates.

**Comment by Zarel** [comment]
Also: do you think users would find "every second" distracting?

I'm kind of considering "every 5 seconds".

File: js/client-battle.js (hunk 4 above)
Comment (Zarel) [inline js/client-battle.js]: The timeout should also be cleared and reset every time the timer is manually updated. This makes it more resilient to network errors.

**Comment by panpawn** [comment]
I really like this idea!

>Also: do you think users would find "every second" distracting?

imo I don't, as long as it's not like obnoxious and flashing or something like that.

**Comment by urkerab** [comment]
I found a bug in `updateControls` - as the last turn plays out, players get access to the "Switch Sides" button, which confuses the UI, e.g. you can fool it into rematching yourself. As I was touching that code anyway for this version, I fixed that at the same time.

(Can you believe I almost wrote `if (this.battle.kickingInactive > ticking)`... yes it does work, since it evaluates to false if either is undefined.)

File: js/client-battle.js [hunk 5]
```diff
@@ -421,8 +424,21 @@
...
+			if ($timerButton.length) {
+				$timerButton.replaceWith(this.getTimerHTML());
+			} else {
+				this.clearInterval(this.timerInterval);
+				this.timerInterval = 0;
+			}
+		},
+		createTimer: function () {
+			if (this.timerInterval) clearInterval(this.timerInterval);
+			this.timerInterval = setInterval(this.updateTimer.bind(this, true), 1000);
```
Comment (Zarel) [inline js/client-battle.js]: `bind` is deprecated, please don't use it

Comment (Zarel) [inline js/client-battle.js]: Also, put all this in `getTimerHTML`, there's no reason to add yet another function to the stack for something this simple.

File: js/client-battle.js [hunk 6]
```diff
@@ -403,9 +406,27 @@
 				break;
 			}
 		},
-		getTimerHTML: function () {
+		timerInterval: 0,
+		getTimerHTML: function (ticking) {
 			var time = 'Timer';
 			var timerTicking = (this.battle.kickingInactive && this.request && !this.request.wait && !(this.choice && this.choice.waiting)) ? ' timerbutton-on' : '';
+
+			if (!ticking) {
```
Comment (Zarel) [inline js/client-battle.js]: um, `timerTicking` already exists; I hope you're aware?

Comment (urkerab) [inline js/client-battle.js]: Bad naming maybe; there were other bugs anyway so I renamed it while I was there.

**Comment by Zarel** [comment]
Okay, looks good.

There's still a style issue but I'll fix that myself, lgtm

**Comment by Zarel** [comment]
Honestly, though, the solution to the `switchSides` issue is to make the rematch button smarter, not to prevent the player from seeing the battle from the other side...

**Pull request merged**